javalib/src/main/scala/java/lang/flttodec/RyuDoubleTables.scala

```diff
@@ -0,0 +1,720 @@
+ *   ((e * 2.321928094887362) + 0.0).toInt + 1
+ *
+ * def formatArray(arr: Array[(Int, Int, Int, Int)], name: String): String = {
+ *   val sb = new StringBuilder
+ *   sb.append(s"val $name = Array(\n")
+ *   for (i <- arr.indices) {
+ *     val (a, b, c, d) = arr(i)
+ *     sb.append(f"  ($a%10d, $b%10d, $c%10d, $d%10d)")
+ *     if (i < arr.length - 1) sb.append(",")
+ *     sb.append("\n")
+ *   }
+ *   sb.append(")\n")
+ *   sb.toString
+ * }
+ *
+ * @main def main(): Unit =
+ *   val mask = BigInteger.valueOf(1).shiftLeft(Pow5QuarterBitCount).subtract(BigInteger.ONE)
+ *   val invMask = BigInteger.valueOf(1).shiftLeft(Pow5InvQuarterBitCount).subtract(BigInteger.ONE)
+ *
+ *   for (i <- 0 until PosTableSize) {
+ *     val pow = BigInteger.valueOf(5).pow(i)
+ *     val pow5len = pow.bitLength()
+ *     val expectedPow5Bits = pow5bits(i)
+ *     if (i < Pow5Split.length) {
+ *       Pow5Split(i) = (
+ *         pow.shiftRight(pow5len - Pow5BitCount + 3 * Pow5QuarterBitCount).and(mask).intValueExact(),
+ *         pow.shiftRight(pow5len - Pow5BitCount + 2 * Pow5QuarterBitCount).and(mask).intValueExact(),
+ *         pow.shiftRight(pow5len - Pow5BitCount + 1 * Pow5QuarterBitCount).and(mask).intValueExact(),
+ *         pow.shiftRight(pow5len - Pow5BitCount + 0 * Pow5QuarterBitCount).and(mask).intValueExact()
+ *       )
+ *     }
+ *     if (i < Pow5InvSplit.length) {
+ *       val j = pow5len - 1 + Pow5InvBitCount
+ *       val inv = BigInteger.ONE.shiftLeft(j).divide(pow).add(BigInteger.ONE)
+ *       Pow5InvSplit(i) = (
+ *         inv.shiftRight(3 * Pow5InvQuarterBitCount).intValueExact(),
+ *         inv.shiftRight(2 * Pow5InvQuarterBitCount).and(invMask).intValueExact(),
+ *         inv.shiftRight(1 * Pow5InvQuarterBitCount).and(invMask).intValueExact(),
+ *         inv.shiftRight(0 * Pow5InvQuarterBitCount).and(invMask).intValueExact()
+ *       )
+ *     }
+ *   }
+ *
+ *   println(s"val PosTableSize = $PosTableSize")
+ *   println(s"val NegTableSize = $NegTableSize")
+ *   println(s"val Pow5BitCount = $Pow5BitCount")
+ *   println(s"val Pow5QuarterBitCount = $Pow5QuarterBitCount")
+ *   println(s"val Pow5InvBitCount = $Pow5InvBitCount")
+ *   println(s"val Pow5InvQuarterBitCount = $Pow5InvQuarterBitCount")
+ *   println()
+ *   println("/** Pow5Split contains pre-computed 5^q values split into 4 31-bit parts. */")
+ *   println(formatArray(Pow5Split, "Pow5Split"))
+ *   println()
+ *   println("/** Pow5InvSplit contains pre-computed floor(2^k/5^q) + 1 values split into 4 31-bit parts. */")
+ *   println(formatArray(Pow5InvSplit, "Pow5InvSplit"))
+ * ```
+ */
+private[flttodec] object RyuDoubleTables {
+  val PosTableSize = 326
+  val NegTableSize = 291
+  val Pow5BitCount = 121
+  val Pow5QuarterBitCount = 31
+  val Pow5InvBitCount = 122
+  val Pow5InvQuarterBitCount = 31
+
+  /** Pow5Split contains pre-computed 5^q values split into 4 31-bit parts. */
+  val Pow5Split = Array(
+    ( 134217728,          0,          0,          0),
```

Comment:
This boxes all the ints. We don't want that here.

We could define a `class FourInts` with 4 `Int` fields (anyway you won't have `Tuple`s available in the linker-private-library). But even that will create a lot of unnecessary code in the generated Wasm.

Consider using a single, flat `Array[Int]`, and index into with `index * 4 + part`.